baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousRecordArrayField.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+
+/**
+ * Warns that users should not have a {@link java.util.regex.Pattern} as a key to a Set or Map.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+        summary = "Record type has an array field and hasn't overridden equals/hashcode. By default array equality"
+                + " performs reference equality only. Consider using an immutable Collection for the field, using"
+                + " Immutables instead of the record, or overriding equals/hashCode in the record.",
+        severity = SeverityLevel.WARNING)
+public final class DangerousRecordArrayField extends BugChecker implements BugChecker.ClassTreeMatcher {
+
+    private static final Matcher<VariableTree> IS_ARRAY_VARIABLE = Matchers.isArrayType();
+    private static final Matcher<MethodTree> NON_TRIVIAL_EQUALS = Matchers.allOf(
+            Matchers.equalsMethodDeclaration(),
+            Matchers.not(Matchers.singleStatementReturnMatcher(Matchers.instanceEqualsInvocation())));
+    private static final Matcher<MethodTree> NON_TRIVIAL_HASHCODE = Matchers.allOf(
+            Matchers.hashCodeMethodDeclaration(),
+            Matchers.not(Matchers.singleStatementReturnMatcher(Matchers.instanceHashCodeInvocation())));
```

Comment:
I don't think we have test coverage for `Matchers.not(Matchers.singleStatementReturnMatcher(Matchers.instanceHashCodeInvocation()))`. Fwiw I think we may be better off searching for both equals and hashCode implementations, there are enough edge cases in how those may be implemented that existence alone should disarm this check.

Reply:
The issue is that `record` generates an `equals` & `hashCode` method. We can't actually check for presence of the method in code (at least, to my knowledge/from my poking while making this PR).

Because of this we also have test coverage btw, the reason `testArray_withEqualsNoHashcode` fails is this bit of the check.

Reply:
I imagine the ClassTree can be searched for method overrides, but the ClassSymbol would include the implicit implementations

Reply:
Where would you expect that? For example,
```
        System.out.println("For " + classSymbol.getSimpleName());
        System.out.println(" Enclosed elements: "
                + classSymbol.getEnclosedElements().stream()
                        .map(Symbol::flatName)
                        .collect(Collectors.toSet()));
```
results in:
```
For MyRecordWithEquals
 Enclosed elements: [payload, toString, hashCode, name, <init>, equals]
For MyRecordWithNone
 Enclosed elements: [payload, toString, hashCode, name, <init>, equals]
For MyRecordWithNone
 Enclosed elements: [payload, toString, hashCode, name, <init>, equals]
For MyRecordWithHashCode
 Enclosed elements: [payload, toString, hashCode, name, <init>, equals]
```

Note that all of them show up identically, despite all of them missing at least one override.

Reply:
I think you want to check `classTree.getMembers()`, but I haven't had a chance to verify this

Reply:
It's possible the symbol could also work if the `Flags.SYNTHETIC` modifier is set on the method.

Reply:
Yes, this actually does differentiate between the two:
```
        System.out.println(" Members: "
                + classTree.getMembers().stream()
                        .filter(tree -> tree instanceof JCTree.JCMethodDecl)
                        .map(tree -> (JCTree.JCMethodDecl) tree)
                        .map(method -> method.name)
                        .collect(Collectors.toSet()));
                        ```

Reply:
Interesting, I had already changed it to using that API, but I think as I went through multiple refactors going from a working to a clean solution I left in a check that was necessary when I was using the ClassSymbol related APIs.